File: frontend/nextjs-site/src/types/lab.ts
export type LabFramework = 'streamlit' | 'gradio'

export type LabType =
  | 'visualization'
  | 'parameter-tuning'
  | 'training-demo'
  | 'dataset-experiment'

export interface Lab {
  id: number
  algorithm_id: number
  lab_name: string
  lab_type: LabType
  url: string | null
  framework: LabFramework
}

File: backend/fastapi-server/app/models/lab.py
from typing import Optional, TYPE_CHECKING
from sqlmodel import SQLModel, Field, Relationship

if TYPE_CHECKING:
    from .algorithm import Algorithm


class Lab(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    algorithm_id: int = Field(foreign_key="algorithm.id", unique=True, index=True)
    lab_name: str
    lab_type: str = Field(default="visualization")
    # "visualization" | "parameter-tuning" | "training-demo" | "dataset-experiment"
    url: Optional[str] = None
    framework: str = Field(default="streamlit")  # "streamlit" | "gradio"
    algorithm: Optional["Algorithm"] = Relationship(back_populates="lab")
